### PR TITLE
clusterloader2: import k8s.io/component-base/metrics/testutil.Histogram to compute quantile

### PR DESCRIPTION
**Pull request by ingvagabund**
On top of https://github.com/kubernetes/perf-tests/pull/978

- importing `k8s.io/component-base/metrics/testutil.Histogram` to compute quantile
- adding unit test to validate the quantile of `k8s.io/perf-tests/clusterloader2/pkg/measurement/util.Histogram` is computed correctly

### CODE REVIEW

**Comment by mm4tt** [comment]
/lgtm

Looks good, will check the cl2 test results before approving

**Comment by ingvagabund** [comment]
@mm4tt feel free to lift the hold, just notice the PR got auto-approved after you lgtm'ed

File: clusterloader2/pkg/measurement/util/histogram.go [hunk 1]
```diff
@@ -28,6 +34,34 @@ type Histogram struct {
...
 
+// Quantile calculates the quantile 'q' based on the given buckets of Histogram.
+func (h *Histogram) Quantile(q float64) (float64, error) {
+	hist := testutil.Histogram{
+		&dto.Histogram{
+			Bucket: []*dto.Bucket{},
+		},
+	}
+
+	for k, v := range h.Buckets {
```
Comment (ahg-g) [inline clusterloader2/pkg/measurement/util/histogram.go]: please add a comment clarifying that hist.Quantile expected a slice of Buckets ordered by UpperBound.

Comment (ingvagabund) [reply clusterloader2/pkg/measurement/util/histogram.go]: Done, right before calling sort.Slice

**Comment by mm4tt** [comment]
Something seems off,  `e2eSchedulingLatency` is being reported as a negative number:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1229767138370129920/artifacts/SchedulingMetrics_load_2020-02-18T14:23:50Z.json

Could you take a look?

**Comment by ingvagabund** [comment]
> Could you take a look?

lemme see

**Comment by ingvagabund** [comment]
@yiyang5055 the histogram built up from the samples is empty. That's why all the quantiles are `NaN` (resp. NaN converted into int64). Did you have a chance to actually test your changes?

**Comment by ingvagabund** [comment]
Based on https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1230076412384251904/artifacts/SchedulingMetrics_load_2020-02-19T10:53:00Z.json there's supposed to be 16 samples. Wondering why there are not added to the histogram.

**Comment by ingvagabund** [comment]
@yiyang5055 `e2eSchedulingMetricName` just needed to be prefixed with `scheduler_`. The histogram is populated correctly now.

**Comment by ingvagabund** [comment]
All good now.

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1230141467285000192/artifacts/SchedulingMetrics_load_2020-02-19T15:17:05Z.json:
```
{
  "predicateEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "priorityEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "e2eSchedulingLatency": {
    "Perc50": 60473988,
    "Perc90": 335236626,
    "Perc99": 788254591
  }
}
```

**Comment by ahg-g** [comment]
/hold
why the other metrics are all zeros?

**Comment by ingvagabund** [comment]
> why the other metrics are all zeros?
let's try once again

**Comment by ahg-g** [comment]
> > why the other metrics are all zeros?
> 
> let's try once again

here is the PR history: https://prow.k8s.io/pr-history/?org=kubernetes&repo=perf-tests&pr=1061

The first three runs (from the right) had those metrics populated, the rest don't.

**Comment by ingvagabund** [comment]
> The first three runs (from the right) had those metrics populated, the rest don't.

Will investigate. Though, I added only few new fields. Weird.

**Comment by ingvagabund** [comment]
@mm4tt pull-perf-tests-clusterloader2 seems to be busted

**Comment by ingvagabund** [comment]
Even here: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1064/pull-perf-tests-clusterloader2/1230826856261881856/artifacts/SchedulingMetrics_load_2020-02-21T12:34:08Z.json from https://github.com/kubernetes/perf-tests/pull/1064:
```json
{
  "predicateEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "priorityEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "e2eSchedulingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  }
}
```

**Comment by ahg-g** [comment]
right, you can see here as well: http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=predicate_evaluation

**Comment by ahg-g** [comment]
/hold

we are probably now looking at deprecated metrics

**Comment by mm4tt** [comment]
Interesting, it seems that it stopped working for master and 1.18 branch (which is daily rebased to master branch)
cQAHfr3oK6P

Our older release tests are fine, e.g. 1.17:
MR3W4CsfkiA

It looks like [these metrics are deprecated since 1.18](https://github.com/kubernetes/kubernetes/blob/d221d82eaf3756a759530bdd332af9ba755151dd/pkg/scheduler/metrics/metrics.go#L76), but they should be still populated, right? 

We can try debugging this, i.e. checking whether this is a scheduler issue (not-populating the metrics) or cl2 issue (not scraping them correctly), but given that they are deprecated, wouldn't be better to just switch to the successor metrics?

Anyway, I opened an issue for this - https://github.com/kubernetes/perf-tests/issues/1074

**Comment by ingvagabund** [comment]
@ahg-g as from https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1076/pull-perf-tests-clusterloader2/1232004003802386432/artifacts/SchedulingMetrics_load_2020-02-24T18:34:12Z.json the metrics file looks like https://gist.github.com/ingvagabund/a9fc18c4ec0ddcfbf2aed84249aa447e.

```
$ cat FILE | grep scheduler_ | grep -v "# " | cut -d'{' -f1 | cut -d' ' -f1 | sort -u
scheduler_binding_duration_seconds_bucket
scheduler_binding_duration_seconds_count
scheduler_binding_duration_seconds_sum
scheduler_e2e_scheduling_duration_seconds_bucket
scheduler_e2e_scheduling_duration_seconds_count
scheduler_e2e_scheduling_duration_seconds_sum
scheduler_framework_extension_point_duration_seconds_bucket
scheduler_framework_extension_point_duration_seconds_count
scheduler_framework_extension_point_duration_seconds_sum
scheduler_pending_pods
scheduler_plugin_execution_duration_seconds_bucket
scheduler_plugin_execution_duration_seconds_count
scheduler_plugin_execution_duration_seconds_sum
scheduler_pod_preemption_victims_bucket
scheduler_pod_preemption_victims_count
scheduler_pod_preemption_victims_sum
scheduler_pod_scheduling_attempts_bucket
scheduler_pod_scheduling_attempts_count
scheduler_pod_scheduling_attempts_sum
scheduler_pod_scheduling_duration_seconds_bucket
scheduler_pod_scheduling_duration_seconds_count
scheduler_pod_scheduling_duration_seconds_sum
scheduler_queue_incoming_pods_total
scheduler_schedule_attempts_total
scheduler_scheduler_cache_size
scheduler_scheduler_goroutines
scheduler_scheduling_algorithm_duration_seconds_bucket
scheduler_scheduling_algorithm_duration_seconds_count
scheduler_scheduling_algorithm_duration_seconds_sum
scheduler_scheduling_algorithm_preemption_evaluation_seconds_bucket
scheduler_scheduling_algorithm_preemption_evaluation_seconds_count
scheduler_scheduling_algorithm_preemption_evaluation_seconds_sum
scheduler_total_preemption_attempts
scheduler_volume_scheduling_duration_seconds_bucket
scheduler_volume_scheduling_duration_seconds_count
scheduler_volume_scheduling_duration_seconds_sum
scheduler_volume_scheduling_stage_error_total
```

None of `scheduling_algorithm_predicate_evaluation_seconds` and `scheduling_algorithm_priority_evaluation_seconds` are reported. Only `binding_duration_seconds` is.

**Comment by ingvagabund** [comment]
Due to https://github.com/kubernetes/kubernetes/pull/86584, both predicate/priority metrics are deprecated since 1.18:
- scheduling_algorithm_predicate_evaluation_seconds, replaced by framework_extension_point_duration_seconds[extenstion_point="Filter"]
- scheduling_algorithm_priority_evaluation_seconds, replaced by framework_extension_point_duration_seconds[extenstion_point="Score"]

**Comment by mm4tt** [comment]
Thanks, @ingvagabund 

In that case, this change looks good to me.
/lgtm

**Comment by ingvagabund** [comment]
Metrics are back!!! From https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1232251530732638208/artifacts/SchedulingMetrics_load_2020-02-25T10:56:11Z.json:
```
{
  "predicateEvaluationLatency": {
    "Perc50": 1303410,
    "Perc90": 6076030,
    "Perc99": 18711003
  },
  "priorityEvaluationLatency": {
    "Perc50": 3324628,
    "Perc90": 12249077,
    "Perc99": 31812986
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 32600018,
    "Perc90": 216257430,
    "Perc99": 463201886
  },
  "e2eSchedulingLatency": {
    "Perc50": 80327804,
    "Perc90": 406599501,
    "Perc99": 873739999
  }
}
```

**Comment by ingvagabund** [comment]
@mm4tt metrics are popping back: http://perf-dash.k8s.io/#/?jobname=gce-100Nodes-master&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=predicate_evaluation

**Pull request merged**